cmd/snap-confine/snap-confine-args.c

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "snap-confine-args.h"
+
+#include <string.h>
+
+#include "../libsnap-confine-private/utils.h"
+
+struct sc_args {
+	// The security tag that the application is intended to run with
+	char *security_tag;
+	// The executable that should be invoked
+	char *executable;
+
+	// Flag indicating that --version was passed on command line.
+	bool is_version_query;
+	// Flag indicating that --classic was passed on command line.
+	bool is_classic_confinement;
+};
+
+struct sc_args *sc_nonfatal_parse_args(int *argcp, char ***argvp,
+				       struct sc_error **errorp)
+{
+	struct sc_args *args = NULL;
+	struct sc_error *err = NULL;
+
+	if (argcp == NULL || argvp == NULL) {
+		err = sc_error_init(SC_ARGS_DOMAIN, 0,
+				    "cannot parse arguments, argcp or argvp is NULL");
+		goto out;
+	}
+	// Use dereferenced versions of argcp and argvp for convenience.
+	int argc = *argcp;
+	char **const argv = *argvp;
+
+	if (argc == 0 || argv == NULL) {
+		err = sc_error_init(SC_ARGS_DOMAIN, 0,
+				    "cannot parse arguments, argc is zero or argv is NULL");
+		goto out;
+	}
+	// Sanity check, look for NULL argv entries.
+	for (int i = 0; i < argc; ++i) {
+		if (argv[i] == NULL) {
+			err = sc_error_init(SC_ARGS_DOMAIN, 0,
+					    "cannot parse arguments, argument at index %d is NULL",
+					    i);
+			goto out;
+		}
+	}
+
+	args = calloc(1, sizeof *args);
+	if (args == NULL) {
+		die("cannot allocate memory for command line arguments object");
+	}
+	// Check if we're being called through the ubuntu-core-launcher symlink.
+	// When this happens we want to skip the first positional argument as it is
+	// the security tag repeated (legacy).
+	bool ignore_first_tag = false;
+	char *basename = strrchr(argv[0], '/');
+	if (basename != NULL) {
+		// NOTE: this is safe because we, at most, may move to the NUL byte
+		// that compares to an empty string.
+		basename += 1;
+		if (strcmp(basename, "ubuntu-core-launcher") == 0) {
+			ignore_first_tag = true;
+		}
+	}
+	// Parse option switches.
```

Comment:
Ok. I think libc's built in parser is likely to be secure and I find it simple to use but I can appreciate the desire to shed some of the unneeded features that it has.